File: Max Number of K-Sum Pairs/max-number-of-k-sum-pairs.py
class Solution(object):
    def maxOperations(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        freq_map = {}
        count = 0
        
        for num in nums:
            diff = k - num
            
            if freq_map.get(diff, 0) > 0:
                freq_map[diff] -= 1
                count += 1
            else:
                freq_map[num] = freq_map.get(num, 0) + 1
        
        return count

File: Max Number of K-Sum Pairs/max-number-of-k-sum-pairs.java
import java.util.HashMap;

class Solution {
    public int maxOperations(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int count = 0;

        for (int num : nums) {
            int diff = k - num;

            if (map.getOrDefault(diff, 0) > 0) {
                map.put(diff, map.get(diff) - 1);
                count++;
            } else {
                map.put(num, map.getOrDefault(num, 0) + 1);
            }
        }

        return count;
    }
}
